### DISCUSSION TITLE
Wayland, Uinput and Autokey

### DISCUSSION THREAD

**Discussion by sebastiansam55** [autokey/autokey#598] [Ideas]
Going forward, Ubuntu (probably the most popular OS for our users) has decided to switch to wayland, this obviously causes a lot of issues for autokey, there are a few things that we should do right now to mitigate this;

1. We should **update the readme to clarify that Ubuntu 21.04 defaults to using wayland** and that it's pretty easy to swap back to X11 as of now (select "Ubuntu" instead of "Ubuntu on Wayland" at the login screen)
2. **Think about how (or even if) we are going to try and support Wayland**. As it stands right now reliance on X11 exists in many files so the changes here are probably going to be quite large.
Another large issue that wayland presents is that I believe a design decision has been made to not have feature parity with X11 especially when it comes to input for security reasons.
3. Someone should begin work on implementing `uinput` as this will workaround some of the wayland limitations. The main question is how should this be implemented, as a complete replacement for the `keyboard` module, a new module `uinput` or changes to the backend to make it an option to swap from X to Uinput for reading/sending events. 

In writing this out it seems like we are going to end up doing a total replacement, as the lesser implementations don't address that Autokey uses X11 for listening for hotkeys.

**Comment by sebastiansam55** [comment]
And obviously not having X11 options removes a lot of other features which cannot be replaced with uinput. Stuff like window filters, most of the `window` module unless `wmctrl` is updated for wayland (it may be already I'm not sure), other stuff I'm sure is dependent on X11

**Reply by josephj11** [reply]
We have a very old issue open for this: #87.

See also PR #844 which points to a repo where uinput was implemented and ydotool which does the same thing using  daemon.

**Comment by josephj11** [comment]
Would it be possible to add an abstraction layer for API calls, etc. that insulates the main code and user scripts from this? Then AutoKey could detect whether X11 is available or not and act appropriately without disturbing the rest of the logic or user scripts.

Would it be possible to write an i/o daemon for AutoKey that could run with elevated permissions and just do whatever lower level i/o requests are needed via API calls? This might get around the restrictions that Wayland imposes on user tasks without opening a huge security vulnerability.

**Reply by tblancher** [reply]
Revisiting this after nearly three years.  About the only thing that's keeping me from Wayland is its inherent lack of support for desktop automation.  It's built deep into the early design of Wayland security.

When I really dove into this sometime last year or earlier this year, it seems Wayland didn't allow for listening for hotkeys and abbreviations (what Keyboard Maestro on macOS calls typed string triggers), which is how I use autokey in the first place.  I may have time over the Christmas 2024 break to look at it again, but I'm not 100 percent sure about that.

I get it, allowing such unprivileged userspace programs like autokey to listen and react to keystrokes is just a half-step removed from keyloggers.  But just as any tool can be used for good or evil, so can desktop automation tools.

This is one place I'd like to help, but I'm not sure where to begin.  `uinput` sounds promising.  The main thing is having time to dig into this.  I'm a working man with a demanding job, a wife who needs my help, and a toddler I'm trying to raise.  Finding time to get on my computer for extra curricular stuff has been a challenge.

**Reply by josephj11** [reply]
Great. Whatever you can do will be appreciated. Family  comes first.

Start by reviewing the develop branch. @sebastiansam55 merged his Wayland changes into it which should handle the keyboard and mouse API calls. I'm not sure about the window and other API calls. Sam said his code works, but we need to test it.

(The "High Level" API calls can just be disabled, not removed. Nobody uses it and the xAutomation project it uses has been  abandoned for quite a while. The x11 version is reported as broken, but may be easy to fix by just fixing a parameter or two in the caling sequences.)

The merge was a bit premature and somehow disabled x11 functionality. Normally Sam would fix this, but he's been absent a long time again. The code should detect the display manager and use the appropriate API methods. I think that's coded, but not working ATM.

There's  a small, but growing Opire bounty for this fix. It's linked to the  actual issue.

I would be supportive of merging a partial fix if everything can't be done now. Having AutoKey partially working under Wayland would be worthwhile - especially if we add a warning about what's not implemented yet so we don't get a ton of unnecessary bug reports.

**Comment by Elliria** [comment]
I thought I'd add that @sebastiansam55 provided [instructions for testing the new AutoKey Wayland code](https://github.com/autokey/autokey/blob/develop/wayland_install.md).
